rfd/0149-headless-kube.md

```diff
@@ -0,0 +1,257 @@
+    Note over Headless: kubectl get pods
+```
+## UX
+User experience while using the headless Kubernetes access will look like this:
+
+```console
+$ tsh --headless proxy kube main_kube_cluster --proxy=tele.port --user=admin
+Complete headless authentication in your local web browser:
+
+https://tele.port:443/web/headless/b1d4215f-a481-5c02-8425-7716b840c62f
+
+or execute this command in your local terminal:
+
+tsh headless approve --user=admin --proxy=tele.port:443 b1d4215f-a481-5c02-8425-7716b840c62f
+
+*** User approves headless request on the local machine ***
+
+Preparing the following Teleport Kubernetes clusters:
+Teleport Cluster Name Kube Cluster Name
+--------------------- -----------------
+tele.port             main_kube_cluster
+
+Started local proxy for Kubernetes on 127.0.0.1:1234 in the background
+and kubectl is set up to work with it. Try issuing a commnd, for example "kubectl get namespaces"
+
+$ kubectl get pods
+NAMESPACE        NAME                                READY   STATUS    RESTARTS      AGE
+teleport-agent   teleport-agent-0                    1/1     Running   0             2d12h
+
+$ exit
+Local Kube proxy is closed.
+
+$ tsh status
+ERROR: Not logged in.
+```
+
+# Security
+
+We will keep the same level of security already achieved by the headless login and Kubernetes local proxy.
+All security-sensitive artifacts created by the local proxy on the remote machine will be kept in memory,
+including private keys and certificates. Additionally, memory is locked from being swapped to the
+disk [`unix.Mlockall`](https://pkg.go.dev/golang.org/x/sys/unix#Mlockall)
+(it was added for the headless login).
+
+We will not change any part of the headless login itself, only integrate the existing
+flow into the `tsh proxy kube` command as an option. Certificates for accessing Kubernetes clusters in Teleport will be
+limited by `max_session_ttl`, which is the same behavior when using the regular `tsh proxy kube`. 
+
+Keeping everything in memory
+and reexecing into a new shell provides a secure and convenient workflow for the user - once the process for the headless kube proxy
+terminates, everything is automatically cleaned up; there are no residual security-sensitive files left on disk.
+It also protects the kube proxy from being accessed by other unprivileged users logged in to the same machine, since they can't access
+the kubeconfig - file location for the kubeconfig `/proce/self/fd/3` can only be read by the owner user, so if another user
+will try to point `kubectl` to it, this wouldn't work. Clients logged in as the same Linux user on this machine will be able to read
+each other's kubeconfig though, therefore if an organization is using a shared remote machine for multiple users, they should make sure
+every person has their own dedicated Linux user login. 
```

Comment:
> Clients logged in as the same Linux user on this machine will be able to read each other's kubeconfig though, therefore if an organization is using a shared remote machine for multiple users, they should make sure every person has their own dedicated Linux user login. 

Do we know anything how often this (i.e. multiple actors using a single Linux login) might be a problem in clients' configurations?